contracts/staking/EpochSelector.sol

```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "./interfaces/IEpochSelector.sol";
+import "./ClusterSelector_wih_abi_encode.sol";
+
+/// @title Contract to select the top 5 clusters in an epoch
+contract EpochSelector is AccessControl, ClusterSelector, IEpochSelector {
+    /// @notice Event emitted when Cluster is selected
+    /// @param epoch Number of Epoch
+    /// @param cluster Address of cluster
+    event ClusterSelected(uint256 indexed epoch, address indexed cluster);
+
+    /// @notice length of epoch
+    uint256 public constant epochLength = 4 hours;
+
+    /// @notice timestamp when the selector starts
+    uint256 public immutable startTime;
+
+    /// @notice Number of clusters selected in every epoch
+    uint256 public immutable numberOfClustersToSelect;
+
+    /// @notice clusters selected during each epoch
+    mapping(uint256 => address[]) public clustersSelected;
+
+    /// @notice ID for update role
+    bytes32 public constant updaterRole = keccak256(abi.encode("updater")); // find standard format for this
+
+    /// @notice ID for updater admin role
+    bytes32 public constant updaterAdminRole = keccak256(abi.encode("updater admin")); // find standard format for this
+
+    constructor(address _admin, uint256 _numberOfClustersToSelect) ClusterSelector() {
+        startTime = block.timestamp;
+        numberOfClustersToSelect = _numberOfClustersToSelect;
+        AccessControl._setRoleAdmin(updaterRole, updaterAdminRole);
+        AccessControl._grantRole(updaterAdminRole, _admin);
+    }
+
+    /// @notice Current Epoch
+    function getCurrentEpoch() public view override returns (uint256) {
+        return (block.timestamp - startTime) / epochLength;
+    }
+
+    /// @notice Returns the list of selected clusters in the current epoch
+    /// @return List of the clusters selected
+    function getCurrentClusters() public override returns (address[] memory) {
+        uint256 epoch = getCurrentEpoch();
+        address[] memory nodes = clustersSelected[epoch];
+        if (nodes.length == 0) {
+            // select and save from the tree
+            clustersSelected[epoch] = selectTopNClusters(block.timestamp, numberOfClustersToSelect);
```

Comment:
If `block.timestamp` is used as randomizer then it means till `getCurrentClusters` function is called  on chain, no one can send any any data on network as it is unknown who the clusters for the epoch are.

Reply:
so it should be possible to set clusters for a future/past epochs before hand. May be hash of block.timestamp and epoch number should do the trick ?

Reply:
We also need to limit how far into future can clusters be selected

Reply:
Let's select clusters for next epoch in current epoch because if we select for too much into future, then selected cluster will be based on delegations at that point. We will give reward to anyone who calls the function that selects clusters. If the cluster selector function is not called then the clusters from prev epoch will be considered.